{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d8aafffd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import glob\n",
    "import random\n",
    "\n",
    "from datetime import datetime\n",
    "from fasta_one_hot_encoder import FastaOneHotEncoder\n",
    "\n",
    "import os\n",
    "import shutil\n",
    "import resource"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06210fd9",
   "metadata": {},
   "source": [
    "## Convert fasta files into one hot encoded files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c61c4fa8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# encoder = FastaOneHotEncoder(\n",
    "#     nucleotides=\"acgt\",\n",
    "#     lower = True,\n",
    "#     sparse = False,\n",
    "#     handle_unknown = \"ignore\")\n",
    "# for fasta in glob.glob(\"../chroms/*.fa\"):\n",
    "#     path = fasta\n",
    "#     chr_tmp = encoder.transform_to_df(path, verbose=True)\n",
    "#     print(path.split(\"/\")[-1].split(\".\")[0])\n",
    "#     chr_tmp.to_csv(\"../chroms/oe_chroms/{}.csv\".format(path.split(\"/\")[-1].split(\".\")[0]))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ff6475a",
   "metadata": {},
   "source": [
    "## Create training data from one hot encoded files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "91b6453f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Before running, increase file descriptor limit\n",
    "# resource.setrlimit(resource.RLIMIT_NOFILE, (131072, 131072))\n",
    "# print (\"getrlimit:\", resource.getrlimit(resource.RLIMIT_NOFILE))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "dba4780a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def createSeqData(chromosomes, step=200, nuc_context=1000):\n",
    "#     pol3_bed_cols_names = [\"Chromosome\", \"Start\", \"End\", \"Name\", \"Score\", \"Strand\"]\n",
    "#     pol3_df = pd.read_csv(\"../data/polr3d.bed\", sep=\"\\s+\", header=None, names=pol3_bed_cols_names)\n",
    "#     rmsk_df = pd.read_csv(\"../data/mm10_rmsk.bed\", sep=\"\\s+\", header=None, names=pol3_bed_cols_names)\n",
    "#     for chrom in chromosomes:\n",
    "#         print(chrom+\":\")\n",
    "#         print(\"     Creating necessary directories...\")\n",
    "#         output_dir1 = \"../data/tmp_seqData/\"\n",
    "#         output_dir2 = \"../data/chr_seqData/\"\n",
    "#         if not os.path.exists(output_dir1):\n",
    "#             os.makedirs(output_dir1)\n",
    "#         if not os.path.exists(output_dir2):\n",
    "#             os.makedirs(output_dir2)\n",
    "            \n",
    "#         #Process chromosome oe df to create training data\n",
    "#         print(\"     Processing one-hot encoded dataframe...\")\n",
    "#         chr_df = pd.read_csv(\"../chroms/oe_chroms/{}.csv\".format(chrom))\n",
    "#         chr_df[\"Label\"] = 0\n",
    "#         pol3_chr_df = pol3_df[pol3_df[\"Chromosome\"] == \"{}\".format(chrom)]\n",
    "#         rmsk_chr_df = rmsk_df[rmsk_df[\"Chromosome\"] == \"{}\".format(chrom)]\n",
    "        \n",
    "#         for row in range(len(rmsk_chr_df)):\n",
    "#             beg_range = rmsk_chr_df.iloc[row][\"Start\"]\n",
    "#             end_range = rmsk_chr_df.iloc[row][\"End\"]\n",
    "#             chr_df.loc[beg_range:end_range, \"Label\"] = 1\n",
    "#         for row in range(len(pol3_chr_df)):\n",
    "#             beg_range = pol3_chr_df.iloc[row][\"Start\"]\n",
    "#             end_range = pol3_chr_df.iloc[row][\"End\"]\n",
    "#             chr_df.loc[beg_range:end_range, \"Label\"] = 2\n",
    "\n",
    "#         print(\"     Creating training data. This may take a while...\")\n",
    "#         #Start creating training data\n",
    "#         labels = []\n",
    "#         file_names = []\n",
    "#         final_data = []\n",
    "#         j = 1\n",
    "#         #Get first and last non-N index\n",
    "#         fasta_sequences = SeqIO.parse(open(\"../chroms/{}.fa\".format(chrom)),'fasta')\n",
    "#         for seq in fasta_sequences:\n",
    "#             name, sequence = seq.id, str(seq.seq)\n",
    "#         a_idx = sequence.lower().index(\"a\")\n",
    "#         c_idx = sequence.lower().index(\"c\")\n",
    "#         g_idx = sequence.lower().index(\"g\")\n",
    "#         t_idx = sequence.lower().index(\"t\")\n",
    "#         chr_start_idx = min(a_idx,c_idx,g_idx,t_idx)\n",
    "#         a_idx = sequence.lower().rfind(\"a\")\n",
    "#         c_idx = sequence.lower().rfind(\"c\")\n",
    "#         g_idx = sequence.lower().rfind(\"g\")\n",
    "#         t_idx = sequence.lower().rfind(\"t\")\n",
    "#         chr_end_idx = max(a_idx,c_idx,g_idx,t_idx)\n",
    "#         for i in range(chr_start_idx, chr_end_idx+1, step):\n",
    "#             if i <= chr_end_idx:\n",
    "#                 beg_seq = []\n",
    "#                 end_seq = []\n",
    "                \n",
    "#                 start_idx = i - nuc_context\n",
    "#                 if start_idx < 0:\n",
    "#                     start_idx = 0\n",
    "#                     n_count = (i - nuc_context) * -1\n",
    "#                     beg_seq = [[0,0,0,0]] * n_count\n",
    "#                 end_idx = i+step+nuc_context\n",
    "#                 if end_idx > len(chr_df):\n",
    "#                     end_idx = len(chr_df)\n",
    "#                     n_count = (i+step+nuc_context) - len(chr_df)\n",
    "#                     end_seq = [[0,0,0,0]] * n_count\n",
    "\n",
    "#                 if beg_seq == [] and end_seq == []:\n",
    "#                     training_seq = chr_df[start_idx:end_idx].drop(columns=[\"Unnamed: 0\", \"Label\"]).to_numpy()\n",
    "#                 elif beg_seq == [] and len(end_seq) != 0:\n",
    "#                     training_seq = chr_df[start_idx:end_idx].drop(columns=[\"Unnamed: 0\", \"Label\"]).to_numpy() + np.array(end_seq)\n",
    "#                 elif len(beg_seq) != 0 and end_seq == []:\n",
    "#                     training_seq = beg_seq + chr_df[start_idx:end_idx].drop(columns=[\"Unnamed: 0\", \"Label\"]).values.tolist()\n",
    "#                 #Determine labels    \n",
    "#                 tmp_df = chr_df[start_idx:end_idx]\n",
    "#                 grouped_df = tmp_df.groupby(\"Label\").count().reset_index()\n",
    "#                 grouped_df.index = grouped_df[\"Label\"]\n",
    "#                 try:\n",
    "#                     if grouped_df[grouped_df[\"Label\"] == 2][\"Unnamed: 0\"][2] >= 65:\n",
    "#                         labels.append([1])\n",
    "#                     elif grouped_df[grouped_df[\"Label\"] == 1][\"Unnamed: 0\"][1] >= 65:\n",
    "#                         labels.append([0])\n",
    "#                     else:\n",
    "#                         labels.append([2])\n",
    "#                 except KeyError:\n",
    "#                     try:\n",
    "#                         if grouped_df[grouped_df[\"Label\"] == 1][\"Unnamed: 0\"][1] >= 65:\n",
    "#                             labels.append([0])\n",
    "#                         else:\n",
    "#                             labels.append([2])\n",
    "#                     except KeyError:\n",
    "#                         labels.append([2])\n",
    "#                 #Save temp files for later concatenation\n",
    "#                 training_seq = np.array([training_seq], dtype=np.uint16)\n",
    "#                 if j == 1: \n",
    "#                     training_data = training_seq\n",
    "#                 else:\n",
    "#                     training_data = np.append(training_data, training_seq, axis=0)\n",
    "#                 if j % 50 == 0:\n",
    "#                     np.savez_compressed(\"../data/tmp_seqData/tmp_{}.npz\".format(i), training_data)\n",
    "#                     file_names.append(\"../data/tmp_seqData/tmp_{}.npz\".format(i))\n",
    "#                     j = 1\n",
    "#                 else:\n",
    "#                     j+=1\n",
    "#         del training_data\n",
    "        \n",
    "#         print(\"     Finalizing training data...\")\n",
    "\n",
    "#         #Open numpy npz files and memmap them to reduce memory usage\n",
    "#         fpath = \"../data/chr_seqData/{}_seqData.dat\".format(chrom)\n",
    "#         rows = 0\n",
    "#         cols = None\n",
    "#         dtype = None\n",
    "#         for data_file in file_names:\n",
    "#             with np.load(data_file) as data:\n",
    "#                 for item in data.files:\n",
    "#                     chunk = data[item]\n",
    "#                     rows += chunk.shape[0]\n",
    "#                     cols = chunk.shape[1]\n",
    "#                     elements = chunk.shape[2]\n",
    "#                     dtype = chunk.dtype\n",
    "                \n",
    "#         merged = np.memmap(fpath, dtype=dtype, mode='w+', shape=(rows, cols, elements))\n",
    "#         idx = 0\n",
    "#         for data_file in file_names:\n",
    "#             with np.load(data_file) as data:\n",
    "#                 for item in data.files:\n",
    "#                     chunk = data[item]\n",
    "#                     merged[idx:idx + len(chunk)] = chunk\n",
    "#                     idx += len(chunk)\n",
    "                \n",
    "#         #Save chr data\n",
    "#         labels = np.array(labels)\n",
    "#         fpath2 = \"../data/chr_seqData/{}_labelsData.npz\".format(chrom)\n",
    "#         np.savez_compressed(fpath2, labels=labels)\n",
    "        \n",
    "#         #Delete temp directories\n",
    "#         dir = '../data/tmp_seqData/'\n",
    "#         shutil.rmtree(dir)\n",
    "        \n",
    "#         print(\"Completed {}!\".format(chrom))\n",
    "#         print(\"Memmap object dimensions:\", rows, columns, elements)\n",
    "        \n",
    "#     print(\"Finished creating training data by chromosome!\")\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9a2eedf5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def createChipData(chromosomes, step=200, nuc_context=1000):\n",
    "#     print(\"Reading in chip files...\")\n",
    "#     chip_dfs = []\n",
    "#     chip_names = []\n",
    "#     chip_bed_cols_names = [\"Chromosome\", \"Start\", \"End\", \"Name\", \"Score\", \"Strand\", \"signalValue\", \"pValue\", \"qValue\", \"peak\"]\n",
    "#     for chip in glob.glob(\"../data/chip_data/*\"):\n",
    "#         tmp_chip_df = pd.read_csv(chip, sep=\"\\s+\", header=None, names=chip_bed_cols_names)\n",
    "#         chip_dfs.append(tmp_chip_df)\n",
    "#         chip_names.append(chip.split(\"/\")[-1].split(\".\")[0])\n",
    "    \n",
    "#     for chrom in chromosomes:\n",
    "#         print(chrom+\":\")\n",
    "#         print(\"     Creating necessary directories...\")\n",
    "#         output_dir1 = \"../data/tmp_chipData/\"\n",
    "#         output_dir2 = \"../data/chr_chipData/\"\n",
    "#         if not os.path.exists(output_dir1):\n",
    "#             os.makedirs(output_dir1)\n",
    "#         if not os.path.exists(output_dir2):\n",
    "#             os.makedirs(output_dir2)\n",
    "            \n",
    "#         #Process chromosome oe df to create training data\n",
    "#         print(\"     Reading in one-hot encoded dataframe...\")\n",
    "#         chr_df = pd.read_csv(\"../chroms/oe_chroms/{}.csv\".format(chrom))\n",
    "            \n",
    "#         #Create ChIP df that is ready to be converted to numpy training data\n",
    "#         print(\"     Processing ChIP data...\")\n",
    "#         chip_df = pd.DataFrame(index=np.arange(len(chr_df))).reset_index().drop(columns=\"index\")\n",
    "#         chip_idx = 0\n",
    "#         for chip in chip_dfs:\n",
    "#             tmp_chip_chr_df = chip[chip[\"Chromosome\"] == \"{}\".format(chrom)]\n",
    "#             chip_df[chip_names[chip_idx]] = 0.\n",
    "            \n",
    "#             for row in range(len(tmp_chip_chr_df)):\n",
    "#                 beg_range = tmp_chip_chr_df.iloc[row][\"Start\"]\n",
    "#                 end_range = tmp_chip_chr_df.iloc[row][\"End\"]\n",
    "#                 chip_df.loc[beg_range:end_range, chip_names[chip_idx]] = tmp_chip_chr_df.iloc[row][\"signalValue\"]\n",
    "#             chip_idx+=1\n",
    "                        \n",
    "\n",
    "#         print(\"     Creating training data. This may take a while...\")\n",
    "#         #Start creating training data\n",
    "#         file_names = []\n",
    "#         final_data = []\n",
    "#         j = 1\n",
    "#         #Get first and last non-N index\n",
    "#         fasta_sequences = SeqIO.parse(open(\"../chroms/{}.fa\".format(chrom)),'fasta')\n",
    "#         for seq in fasta_sequences:\n",
    "#             name, sequence = seq.id, str(seq.seq)\n",
    "#         a_idx = sequence.lower().index(\"a\")\n",
    "#         c_idx = sequence.lower().index(\"c\")\n",
    "#         g_idx = sequence.lower().index(\"g\")\n",
    "#         t_idx = sequence.lower().index(\"t\")\n",
    "#         chr_start_idx = min(a_idx,c_idx,g_idx,t_idx)\n",
    "#         a_idx = sequence.lower().rfind(\"a\")\n",
    "#         c_idx = sequence.lower().rfind(\"c\")\n",
    "#         g_idx = sequence.lower().rfind(\"g\")\n",
    "#         t_idx = sequence.lower().rfind(\"t\")\n",
    "#         chr_end_idx = max(a_idx,c_idx,g_idx,t_idx)\n",
    "#         for i in range(chr_start_idx, chr_end_idx+1, step):\n",
    "#             if i <= chr_end_idx:\n",
    "#                 beg_seq = []\n",
    "#                 end_seq = []\n",
    "                \n",
    "#                 start_idx = i - nuc_context\n",
    "#                 if start_idx < 0:\n",
    "#                     start_idx = 0\n",
    "#                     n_count = (i - nuc_context) * -1\n",
    "#                     beg_seq = [[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]] * n_count\n",
    "#                 end_idx = i+step+nuc_context\n",
    "#                 if end_idx > len(chr_df):\n",
    "#                     end_idx = len(chr_df)\n",
    "#                     n_count = (i+step+nuc_context) - len(chr_df)\n",
    "#                     end_seq = [[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]] * n_count\n",
    "\n",
    "#                 if beg_seq == [] and end_seq == []:\n",
    "#                     training_seq = chip_df[start_idx:end_idx].to_numpy()\n",
    "#                 elif beg_seq == [] and len(end_seq) != 0:\n",
    "#                     training_seq = chip_df[start_idx:end_idx].to_numpy() + np.array(end_seq)\n",
    "#                 elif len(beg_seq) != 0 and end_seq == []:\n",
    "#                     training_seq = beg_seq + chip_df[start_idx:end_idx].values.tolist()\n",
    "\n",
    "#                 #Save temp files for later concatenation\n",
    "#                 training_seq = np.array([training_seq], dtype=np.float32)\n",
    "#                 if j == 1: \n",
    "#                     training_data = training_seq\n",
    "#                 else:\n",
    "#                     training_data = np.append(training_data, training_seq, axis=0)\n",
    "#                 if j % 50 == 0:\n",
    "                    \n",
    "#                     np.savez_compressed(\"../data/tmp_chipData/tmp_{}.npz\".format(i), training_data)\n",
    "#                     file_names.append(\"../data/tmp_chipData/tmp_{}.npz\".format(i))\n",
    "#                     j = 1\n",
    "#                 else:\n",
    "#                     j+=1\n",
    "#         del training_data\n",
    "        \n",
    "#         print(\"     Finalizing training data...\")\n",
    "#         #Open numpy npz files and memmap them to reduce memory usage\n",
    "#         fpath = \"../data/chr_chipData/{}_chipData.dat\".format(chrom)\n",
    "#         rows = 0\n",
    "#         cols = None\n",
    "#         dtype = None\n",
    "#         for data_file in file_names:\n",
    "#             with np.load(data_file) as data:\n",
    "#                 for item in data.files:\n",
    "#                     chunk = data[item]\n",
    "#                     rows += chunk.shape[0]\n",
    "#                     cols = chunk.shape[1]\n",
    "#                     elements = chunk.shape[2]\n",
    "#                     dtype = chunk.dtype\n",
    "                \n",
    "#         merged = np.memmap(fpath, dtype=dtype, mode='w+', shape=(rows, cols, elements))\n",
    "#         idx = 0\n",
    "#         for data_file in file_names:\n",
    "#             with np.load(data_file) as data:\n",
    "#                 for item in data.files:\n",
    "#                     chunk = data[item]\n",
    "#                     merged[idx:idx + len(chunk)] = chunk\n",
    "#                     idx += len(chunk)\n",
    "        \n",
    "#         #Delete temp directories\n",
    "#         dir = '../data/tmp_chipData/'\n",
    "#         shutil.rmtree(dir)\n",
    "        \n",
    "#         print(\"Completed {}!\".format(chrom))\n",
    "#         print(\"Memmap object dimensions:\", rows, columns, elements)\n",
    "        \n",
    "#     print(\"Finished creating training data by chromosome!\")\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fc9f2329",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def modifyLabels(chromosomes, step=200, nuc_context=1000):\n",
    "#     pol3_bed_cols_names = [\"Chromosome\", \"Start\", \"End\", \"Name\", \"Score\", \"Strand\"]\n",
    "#     pol3_df = pd.read_csv(\"../data/polr3d.bed\", sep=\"\\s+\", header=None, names=pol3_bed_cols_names)\n",
    "#     rmsk_df = pd.read_csv(\"../data/mm10_rmsk.bed\", sep=\"\\s+\", header=None, names=pol3_bed_cols_names)\n",
    "#     for chrom in chromosomes:\n",
    "#         print(chrom+\":\")\n",
    "#         print(\"     Creating necessary directories...\")\n",
    "#         output_dir1 = \"../data/tmp_seqData/\"\n",
    "#         output_dir2 = \"../data/chr_seqData/\"\n",
    "#         if not os.path.exists(output_dir1):\n",
    "#             os.makedirs(output_dir1)\n",
    "#         if not os.path.exists(output_dir2):\n",
    "#             os.makedirs(output_dir2)\n",
    "            \n",
    "#         #Process chromosome oe df to create training data\n",
    "#         print(\"     Processing one-hot encoded dataframe...\")\n",
    "#         chr_df = pd.read_csv(\"../chroms/oe_chroms/{}.csv\".format(chrom))\n",
    "#         chr_df[\"Label\"] = 0\n",
    "#         pol3_chr_df = pol3_df[pol3_df[\"Chromosome\"] == \"{}\".format(chrom)]\n",
    "#         rmsk_chr_df = rmsk_df[rmsk_df[\"Chromosome\"] == \"{}\".format(chrom)]\n",
    "        \n",
    "#         for row in range(len(rmsk_chr_df)):\n",
    "#             beg_range = rmsk_chr_df.iloc[row][\"Start\"]\n",
    "#             end_range = rmsk_chr_df.iloc[row][\"End\"]\n",
    "#             chr_df.loc[beg_range:end_range, \"Label\"] = 1\n",
    "#         for row in range(len(pol3_chr_df)):\n",
    "#             beg_range = pol3_chr_df.iloc[row][\"Start\"]\n",
    "#             end_range = pol3_chr_df.iloc[row][\"End\"]\n",
    "#             chr_df.loc[beg_range:end_range, \"Label\"] = 2\n",
    "            \n",
    "\n",
    "#         print(\"     Creating training data. This may take a while...\")\n",
    "#         #Start creating training data\n",
    "#         labels = []\n",
    "#         file_names = []\n",
    "#         final_data = []\n",
    "#         j = 1\n",
    "#         #Get first and last non-N index\n",
    "#         fasta_sequences = SeqIO.parse(open(\"../chroms/{}.fa\".format(chrom)),'fasta')\n",
    "#         for seq in fasta_sequences:\n",
    "#             name, sequence = seq.id, str(seq.seq)\n",
    "#         a_idx = sequence.lower().index(\"a\")\n",
    "#         c_idx = sequence.lower().index(\"c\")\n",
    "#         g_idx = sequence.lower().index(\"g\")\n",
    "#         t_idx = sequence.lower().index(\"t\")\n",
    "#         chr_start_idx = min(a_idx,c_idx,g_idx,t_idx)\n",
    "#         a_idx = sequence.lower().rfind(\"a\")\n",
    "#         c_idx = sequence.lower().rfind(\"c\")\n",
    "#         g_idx = sequence.lower().rfind(\"g\")\n",
    "#         t_idx = sequence.lower().rfind(\"t\")\n",
    "#         chr_end_idx = max(a_idx,c_idx,g_idx,t_idx)\n",
    "#         for i in range(chr_start_idx, chr_end_idx+1, step):\n",
    "#             if i <= chr_end_idx:\n",
    "#                 beg_seq = []\n",
    "#                 end_seq = []\n",
    "                \n",
    "#                 start_idx = i - nuc_context\n",
    "#                 if start_idx < 0:\n",
    "#                     start_idx = 0\n",
    "#                     n_count = (i - nuc_context) * -1\n",
    "#                     beg_seq = [[0,0,0,0]] * n_count\n",
    "#                 end_idx = i+step+nuc_context\n",
    "#                 if end_idx > len(chr_df):\n",
    "#                     end_idx = len(chr_df)\n",
    "#                     n_count = (i+step+nuc_context) - len(chr_df)\n",
    "#                     end_seq = [[0,0,0,0]] * n_count\n",
    "\n",
    "#                 #Determine labels    \n",
    "#                 tmp_df = chr_df[start_idx:end_idx]\n",
    "#                 grouped_df = tmp_df.groupby(\"Label\").count().reset_index()\n",
    "#                 try:\n",
    "#                     if grouped_df[grouped_df[\"Label\"] == 2][\"Unnamed: 0\"][2] >= 65:\n",
    "#                         labels.append([1])\n",
    "#                     elif grouped_df[grouped_df[\"Label\"] == 1][\"Unnamed: 0\"][2] >= 65:\n",
    "#                         labels.append([2])\n",
    "#                     else:\n",
    "#                         labels.append([0])\n",
    "#                 except KeyError:\n",
    "#                     try:\n",
    "#                         if grouped_df[grouped_df[\"Label\"] == 1][\"Unnamed: 0\"][2] >= 65:\n",
    "#                             labels.append([1])\n",
    "#                         else:\n",
    "#                             labels.append([0])\n",
    "#                     except KeyError:\n",
    "#                         labels.append([0])\n",
    "        \n",
    "#         print(\"     Finalizing training data...\")\n",
    "\n",
    "                \n",
    "#         #Save chr data\n",
    "#         labels = np.array(labels)\n",
    "#         fpath2 = \"../data/chr_seqData/{}_labelsData.npz\".format(chrom)\n",
    "#         np.savez_compressed(fpath2, labels=labels)\n",
    "        \n",
    "        \n",
    "#         print(\"Completed {}!\".format(chrom))\n",
    "        \n",
    "#     print(\"Finished modifying labels data by chromosome!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf0b6695",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "856048c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def createPol3Data(chromosomes, dt=\"training\", step=200, nuc_context=1000):\n",
    "    pol3_bed_cols_names = [\"Chromosome\", \"Start\", \"End\", \"Name\", \"Score\", \"Strand\"]\n",
    "    pol3_df = pd.read_csv(\"../data/polr3d.bed\", sep=\"\\s+\", header=None, names=pol3_bed_cols_names)\n",
    "    rmsk_df = pd.read_csv(\"../data/mm10_rmsk.bed\", sep=\"\\s+\", header=None, names=pol3_bed_cols_names)\n",
    "    nucOE_df = pd.DataFrame({\"Nucleotide\": [\"a\", \"c\", \"g\", \"t\", \"n\"], \"OE\": [[1,0,0,0],[0,1,0,0],[0,0,1,0],[0,0,0,1],[0,0,0,0]]})\n",
    "    j = 1\n",
    "    \n",
    "    print(\"Reading in ChIP files...\")\n",
    "    chip_dfs = []\n",
    "    chip_names = []\n",
    "    chip_bed_cols_names = [\"Chromosome\", \"Start\", \"End\", \"Name\", \"Score\", \"Strand\", \"signalValue\", \"pValue\", \"qValue\", \"peak\"]\n",
    "    for chip in glob.glob(\"../data/chip_data/*\"):\n",
    "        tmp_chip_df = pd.read_csv(chip, sep=\"\\s+\", header=None, names=chip_bed_cols_names)\n",
    "        chip_dfs.append(tmp_chip_df)\n",
    "        chip_names.append(chip.split(\"/\")[-1].split(\".\")[0])\n",
    "    del tmp_chip_df\n",
    "        \n",
    "    for chrom in chromosomes:\n",
    "        print(chrom+\":\")\n",
    "        print(\"     Creating necessary directories...\")\n",
    "        output_dir1 = \"../data/mlData/\"\n",
    "        if not os.path.exists(output_dir1):\n",
    "            os.makedirs(output_dir1)\n",
    "            \n",
    "        #Process chromosome oe df to create training data\n",
    "        print(\"     Processing FASTA sequence...\")\n",
    "        fasta_sequences = SeqIO.parse(open(\"../chroms/{}.fa\".format(chrom)),'fasta')\n",
    "        for seq in fasta_sequences:\n",
    "            name, sequence = seq.id, str(seq.seq).lower()\n",
    "        del fasta_sequences\n",
    "        chr_df = pd.DataFrame({\"Nucleotide\": list(sequence)})\n",
    "        chr_df[\"Label\"] = 0\n",
    "        pol3_chr_df = pol3_df[pol3_df[\"Chromosome\"] == \"{}\".format(chrom)]\n",
    "        rmsk_chr_df = rmsk_df[rmsk_df[\"Chromosome\"] == \"{}\".format(chrom)]\n",
    "        \n",
    "        for row in range(len(rmsk_chr_df)):\n",
    "            beg_range = rmsk_chr_df.iloc[row][\"Start\"]\n",
    "            end_range = rmsk_chr_df.iloc[row][\"End\"]\n",
    "            chr_df.loc[beg_range:end_range, \"Label\"] = 1\n",
    "        for row in range(len(pol3_chr_df)):\n",
    "            beg_range = pol3_chr_df.iloc[row][\"Start\"]\n",
    "            end_range = pol3_chr_df.iloc[row][\"End\"]\n",
    "            chr_df.loc[beg_range:end_range, \"Label\"] = 2\n",
    "        del pol3_chr_df\n",
    "        del rmsk_chr_df\n",
    "                        \n",
    "        #Create ChIP df that is ready to be converted to numpy training data\n",
    "        print(\"     Processing ChIP data...\")\n",
    "        chip_df = pd.DataFrame(index=np.arange(len(chr_df))).reset_index().drop(columns=\"index\")\n",
    "        chip_idx = 0\n",
    "        for chip in chip_dfs:\n",
    "            tmp_chip_chr_df = chip[chip[\"Chromosome\"] == \"{}\".format(chrom)]\n",
    "            chip_df[chip_names[chip_idx]] = 0.\n",
    "            \n",
    "            for row in range(len(tmp_chip_chr_df)):\n",
    "                beg_range = tmp_chip_chr_df.iloc[row][\"Start\"]\n",
    "                end_range = tmp_chip_chr_df.iloc[row][\"End\"]\n",
    "                chip_df.loc[beg_range:end_range, chip_names[chip_idx]] = tmp_chip_chr_df.iloc[row][\"signalValue\"]\n",
    "            chip_idx+=1\n",
    "\n",
    "        print(\"     Creating data for model. This may take a while...\")\n",
    "        #Start creating training data\n",
    "        labels = []\n",
    "        #Get first and last non-N index\n",
    "        a_idx = sequence.index(\"a\")\n",
    "        c_idx = sequence.index(\"c\")\n",
    "        g_idx = sequence.index(\"g\")\n",
    "        t_idx = sequence.index(\"t\")\n",
    "        chr_start_idx = min(a_idx,c_idx,g_idx,t_idx)\n",
    "        a_idx = sequence.rfind(\"a\")\n",
    "        c_idx = sequence.rfind(\"c\")\n",
    "        g_idx = sequence.rfind(\"g\")\n",
    "        t_idx = sequence.rfind(\"t\")\n",
    "        chr_end_idx = max(a_idx,c_idx,g_idx,t_idx)\n",
    "        for i in range(chr_start_idx, chr_end_idx+1, step):\n",
    "            if i <= chr_end_idx:\n",
    "                beg_seq = []\n",
    "                end_seq = []\n",
    "                \n",
    "                beg_chip = []\n",
    "                end_chip = []\n",
    "                \n",
    "                start_idx = i - nuc_context\n",
    "                if start_idx < 0:\n",
    "                    start_idx = 0\n",
    "                    n_count = (i - nuc_context) * -1\n",
    "                    beg_seq = [[0,0,0,0]] * n_count\n",
    "                    beg_chip = [[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]] * n_count\n",
    "                end_idx = i+step+nuc_context\n",
    "                if end_idx > len(chr_df):\n",
    "                    end_idx = len(chr_df)\n",
    "                    n_count = (i+step+nuc_context) - len(chr_df)\n",
    "                    end_seq = [[0,0,0,0]] * n_count\n",
    "                    end_chip = [[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]] * n_count\n",
    "                    \n",
    "                #Determine labels    \n",
    "                tmp_df = chr_df[i:i+step]\n",
    "                grouped_df = tmp_df.groupby(\"Label\").count().reset_index()\n",
    "                grouped_df.index = grouped_df[\"Label\"]\n",
    "                try:\n",
    "                    if grouped_df[grouped_df[\"Label\"] == 2][\"Nucleotide\"][2] >= 65:\n",
    "                        label = [1]\n",
    "                    elif grouped_df[grouped_df[\"Label\"] == 1][\"Nucleotide\"][1] >= 65:\n",
    "                        label = [0]\n",
    "                    else:\n",
    "                        label = [2]\n",
    "                except KeyError:\n",
    "                    try:\n",
    "                        if grouped_df[grouped_df[\"Label\"] == 1][\"Nucleotide\"][1] >= 65:\n",
    "                            label = [0]\n",
    "                        else:\n",
    "                            label = [2]\n",
    "                    except KeyError:\n",
    "                        label = [2]\n",
    "                del tmp_df\n",
    "                del grouped_df\n",
    "                        \n",
    "                if label == [0] or label == [1]:\n",
    "                    #Used to randomly select which nucleotide sequences with label [0] are in datasets\n",
    "                    if label == [0]:\n",
    "                        random_num = random.randint(-1000,1)\n",
    "                    if (label == [0] and random_num > 0) or (label == [1]):                             \n",
    "                        curr_chr_df = chr_df[start_idx:end_idx].merge(nucOE_df, on=[\"Nucleotide\"], how=\"left\")\n",
    "\n",
    "                        if beg_seq == [] and end_seq == []:\n",
    "                            curr_seq = curr_chr_df[\"OE\"].tolist()\n",
    "                        elif beg_seq == [] and len(end_seq) != 0:\n",
    "                            curr_seq = curr_chr_df[\"OE\"].tolist() + end_seq\n",
    "                        elif len(beg_seq) != 0 and end_seq == []:\n",
    "                            curr_seq = beg_seq + curr_chr_df[\"OE\"].tolist()\n",
    "                        del curr_chr_df\n",
    "\n",
    "                        if beg_chip == [] and end_chip == []:\n",
    "                            curr_chip = chip_df[start_idx:end_idx].values.tolist()\n",
    "                        elif beg_chip == [] and len(end_chip) != 0:\n",
    "                            curr_chip = chip_df[start_idx:end_idx].values.tolist() + end_chip\n",
    "                        elif len(beg_chip) != 0 and end_chip == []:\n",
    "                            curr_chip = beg_chip + chip_df[start_idx:end_idx].values.tolist()\n",
    "\n",
    "                        #Save current sequences\n",
    "                        curr_seq = np.array([curr_seq], dtype=np.uint16)\n",
    "                        curr_chip = np.array([curr_chip], dtype=np.float64)\n",
    "                        labels.append(label)\n",
    "                        if j == 1: \n",
    "                            seq_data = curr_seq\n",
    "                            chip_data = curr_chip\n",
    "                            j+=1\n",
    "                        else:\n",
    "                            seq_data = np.append(seq_data, curr_seq, axis=0)\n",
    "                            chip_data = np.append(chip_data, curr_chip, axis=0)\n",
    "                        del curr_seq\n",
    "                        del curr_chip\n",
    "\n",
    "        print(\"Completed {}!\".format(chrom))\n",
    "\n",
    "    print(\"Finalizing data...\")\n",
    "    labels = np.array(labels)\n",
    "    true_idx = np.where(labels == [1])[0]\n",
    "    false_idx = np.where(labels == [0])[0]\n",
    "\n",
    "    #Shuffle and subset indeces for training and testing datasets\n",
    "\n",
    "    np.random.shuffle(false_idx)\n",
    "    false_idx = false_idx[:len(true_idx)]\n",
    "    seq_data = np.append(seq_data[true_idx], seq_data[false_idx], axis=0)\n",
    "    chip_data = np.append(chip_data[true_idx], chip_data[false_idx], axis=0)\n",
    "    \n",
    "    labels = np.append(labels[true_idx], labels[false_idx], axis=0)\n",
    "    \n",
    "    final_idx = np.random.permutation(len(labels))\n",
    "    seq_data = seq_data[final_idx]\n",
    "    chip_data = chip_data[final_idx]\n",
    "    labels = labels[final_idx]\n",
    "    \n",
    "    np.savez_compressed(\"../data/mlData/{}_seqData.npz\".format(dt), dna=seq_data, label=labels)\n",
    "    np.savez_compressed(\"../data/mlData/{}_chipData.npz\".format(dt), chip=chip_data, label=labels)\n",
    "        \n",
    "    print(\"Finished creating data!\")\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "10287186",
   "metadata": {},
   "outputs": [],
   "source": [
    "training_chroms = [\"chr1\", \"chr2\", \"chr3\", \"chr4\", \"chr5\", \"chr6\", \"chr9\", \"chr10\", \"chr11\", \"chr12\", \"chr13\", \"chr14\",\n",
    "                  \"chr15\", \"chr16\", \"chr17\", \"chr18\", \"chr19\", \"chr20\", \"chr21\", \"chr22\", \"chrX\", \"chrY\"]\n",
    "testing_chroms = [\"chr7\", \"chr8\"]\n",
    "all_chroms = [\"chr1\", \"chr2\", \"chr3\", \"chr4\", \"chr5\", \"chr6\", \"chr7\", \"chr8\", \"chr9\", \"chr10\", \"chr11\", \"chr12\", \"chr13\", \"chr14\",\n",
    "                  \"chr15\", \"chr16\", \"chr17\", \"chr18\", \"chr19\", \"chr20\", \"chr21\", \"chr22\", \"chrX\", \"chrY\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "716df38e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading in ChIP files...\n",
      "chr7:\n",
      "     Creating necessary directories...\n",
      "     Processing FASTA sequence...\n",
      "     Processing ChIP data...\n",
      "     Creating data for model. This may take a while...\n",
      "Completed chr7!\n",
      "chr8:\n",
      "     Creating necessary directories...\n",
      "     Processing FASTA sequence...\n",
      "     Processing ChIP data...\n",
      "     Creating data for model. This may take a while...\n",
      "Completed chr8!\n",
      "Finalizing data...\n",
      "Finished creating data!\n"
     ]
    }
   ],
   "source": [
    "createPol3Data(testing_chroms, dt=\"testing\", step=200, nuc_context=1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "71e9ff19",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading in ChIP files...\n",
      "chr1:\n",
      "     Creating necessary directories...\n",
      "     Processing FASTA sequence...\n",
      "     Processing ChIP data...\n",
      "     Creating data for model. This may take a while...\n",
      "Completed chr1!\n",
      "chr2:\n",
      "     Creating necessary directories...\n",
      "     Processing FASTA sequence...\n",
      "     Processing ChIP data...\n",
      "     Creating data for model. This may take a while...\n",
      "Completed chr2!\n",
      "chr3:\n",
      "     Creating necessary directories...\n",
      "     Processing FASTA sequence...\n",
      "     Processing ChIP data...\n",
      "     Creating data for model. This may take a while...\n",
      "Completed chr3!\n",
      "chr4:\n",
      "     Creating necessary directories...\n",
      "     Processing FASTA sequence...\n",
      "     Processing ChIP data...\n",
      "     Creating data for model. This may take a while...\n",
      "Completed chr4!\n",
      "chr5:\n",
      "     Creating necessary directories...\n",
      "     Processing FASTA sequence...\n",
      "     Processing ChIP data...\n",
      "     Creating data for model. This may take a while...\n",
      "Completed chr5!\n",
      "chr6:\n",
      "     Creating necessary directories...\n",
      "     Processing FASTA sequence...\n",
      "     Processing ChIP data...\n",
      "     Creating data for model. This may take a while...\n",
      "Completed chr6!\n",
      "chr9:\n",
      "     Creating necessary directories...\n",
      "     Processing FASTA sequence...\n",
      "     Processing ChIP data...\n",
      "     Creating data for model. This may take a while...\n",
      "Completed chr9!\n",
      "chr10:\n",
      "     Creating necessary directories...\n",
      "     Processing FASTA sequence...\n",
      "     Processing ChIP data...\n",
      "     Creating data for model. This may take a while...\n",
      "Completed chr10!\n",
      "chr11:\n",
      "     Creating necessary directories...\n",
      "     Processing FASTA sequence...\n",
      "     Processing ChIP data...\n",
      "     Creating data for model. This may take a while...\n",
      "Completed chr11!\n",
      "chr12:\n",
      "     Creating necessary directories...\n",
      "     Processing FASTA sequence...\n",
      "     Processing ChIP data...\n",
      "     Creating data for model. This may take a while...\n",
      "Completed chr12!\n",
      "chr13:\n",
      "     Creating necessary directories...\n",
      "     Processing FASTA sequence...\n",
      "     Processing ChIP data...\n",
      "     Creating data for model. This may take a while...\n",
      "Completed chr13!\n",
      "chr14:\n",
      "     Creating necessary directories...\n",
      "     Processing FASTA sequence...\n",
      "     Processing ChIP data...\n",
      "     Creating data for model. This may take a while...\n",
      "Completed chr14!\n",
      "chr15:\n",
      "     Creating necessary directories...\n",
      "     Processing FASTA sequence...\n",
      "     Processing ChIP data...\n",
      "     Creating data for model. This may take a while...\n",
      "Completed chr15!\n",
      "chr16:\n",
      "     Creating necessary directories...\n",
      "     Processing FASTA sequence...\n",
      "     Processing ChIP data...\n",
      "     Creating data for model. This may take a while...\n",
      "Completed chr16!\n",
      "chr17:\n",
      "     Creating necessary directories...\n",
      "     Processing FASTA sequence...\n",
      "     Processing ChIP data...\n",
      "     Creating data for model. This may take a while...\n",
      "Completed chr17!\n",
      "chr18:\n",
      "     Creating necessary directories...\n",
      "     Processing FASTA sequence...\n",
      "     Processing ChIP data...\n",
      "     Creating data for model. This may take a while...\n",
      "Completed chr18!\n",
      "chr19:\n",
      "     Creating necessary directories...\n",
      "     Processing FASTA sequence...\n",
      "     Processing ChIP data...\n",
      "     Creating data for model. This may take a while...\n",
      "Completed chr19!\n",
      "chr20:\n",
      "     Creating necessary directories...\n",
      "     Processing FASTA sequence...\n",
      "     Processing ChIP data...\n",
      "     Creating data for model. This may take a while...\n",
      "Completed chr20!\n",
      "chr21:\n",
      "     Creating necessary directories...\n",
      "     Processing FASTA sequence...\n",
      "     Processing ChIP data...\n",
      "     Creating data for model. This may take a while...\n",
      "Completed chr21!\n",
      "chr22:\n",
      "     Creating necessary directories...\n",
      "     Processing FASTA sequence...\n",
      "     Processing ChIP data...\n",
      "     Creating data for model. This may take a while...\n",
      "Completed chr22!\n",
      "chrX:\n",
      "     Creating necessary directories...\n",
      "     Processing FASTA sequence...\n",
      "     Processing ChIP data...\n",
      "     Creating data for model. This may take a while...\n",
      "Completed chrX!\n",
      "chrY:\n",
      "     Creating necessary directories...\n",
      "     Processing FASTA sequence...\n",
      "     Processing ChIP data...\n",
      "     Creating data for model. This may take a while...\n",
      "Completed chrY!\n",
      "Finalizing data...\n",
      "Finished creating data!\n"
     ]
    }
   ],
   "source": [
    "createPol3Data(training_chroms, dt=\"training\", step=200, nuc_context=1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f522613e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# createSeqData(all_chroms, step=200, nuc_context=1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47ac815d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# createChipData(all_chroms, step=200, nuc_context=1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aff1f420",
   "metadata": {},
   "outputs": [],
   "source": [
    "# modifyLabels([\"chr20\", \"chr21\", \"chr22\", \"chrX\", \"chrY\"], step=200, nuc_context=1000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddcb2cb9",
   "metadata": {},
   "source": [
    "## Create final training data from chromosome data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10b791a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# seq_rows = [1244650,1210850,991100,950950,907300,853400,796600,725050,691600,668850,675050,666250,491750,454400,\n",
    "#            424900,451050,415900,401250,292700,321350,208450,201450,780100,286000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b9c729c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# data = \"seq\"\n",
    "# idx = 0\n",
    "# for chrom in all_chroms:\n",
    "#     print(chrom)\n",
    "#     labels = np.load(\"../data/chr_seqData/{}_labelsData.npz\".format(chrom))\n",
    "#     for item in labels.files:\n",
    "#         true_idx = np.where(labels[item] == [1])[0]\n",
    "#         false_idx = np.where(labels[item] == [0])[0]\n",
    "        \n",
    "#     #Shuffle and subset indeces for training and testing datasets\n",
    "#     np.random.shuffle(true_idx)\n",
    "#     training_true_idx = true_idx[:int(len(true_idx)*0.9)]\n",
    "#     testing_true_idx = true_idx[int(len(true_idx)*0.9)+1:]\n",
    "    \n",
    "#     np.random.shuffle(false_idx)\n",
    "#     training_false_idx = false_idx[:int(len(true_idx)*0.9)]\n",
    "#     testing_false_idx = false_idx[int(len(true_idx)*0.9)+1:len(true_idx)]\n",
    "    \n",
    "#     if data == \"seq\":\n",
    "#         dtype = np.uint16\n",
    "#         elements = 4\n",
    "#         name = \"Sequence\"\n",
    "#     elif data == \"chip\":\n",
    "#         dtype = np.float32\n",
    "#         elements = 20\n",
    "#         name = \"Chip\"\n",
    "    \n",
    "#     chr_seq = np.memmap(\"../data/chr_{}Data/{}_{}Data.dat\".format(data,chrom,data), dtype=dtype, mode='r', shape=(seq_rows[idx],2200,elements))\n",
    "#     chr_seq_true_training = chr_seq[training_true_idx]\n",
    "#     chr_seq_true_testing = chr_seq[testing_true_idx]\n",
    "#     chr_seq_false_training = chr_seq[training_false_idx]\n",
    "#     chr_seq_false_testing = chr_seq[testing_false_idx]\n",
    "    \n",
    "#     #Append data to respective matrix\n",
    "#     if idx == 0:\n",
    "#         chr_seq_training = np.append(chr_seq_true_training, chr_seq_false_training, axis=0)\n",
    "#         chr_seq_testing = np.append(chr_seq_true_testing, chr_seq_false_testing, axis=0)\n",
    "#     else:\n",
    "#         chr_seq_training = np.append(chr_seq_training, np.append(chr_seq_true_training, chr_seq_false_training, axis=0), axis=0)\n",
    "#         chr_seq_testing = np.append(chr_seq_testing, np.append(chr_seq_true_testing, chr_seq_false_testing, axis=0), axis=0)\n",
    "        \n",
    "#     #Keep track of labels\n",
    "#     for item in labels.files:\n",
    "#         if idx == 0:\n",
    "#             final_labels_training = labels[item][training_true_idx]\n",
    "#             final_labels_training = np.append(final_labels_training, labels[item][training_false_idx], axis=0)\n",
    "            \n",
    "#             final_labels_testing = labels[item][testing_true_idx]\n",
    "#             final_labels_testing = np.append(final_labels_testing, labels[item][testing_false_idx], axis=0)\n",
    "#         else:\n",
    "#             final_labels_training = np.append(final_labels_training, labels[item][training_true_idx], axis=0)\n",
    "#             final_labels_training = np.append(final_labels_training, labels[item][training_false_idx], axis=0)\n",
    "            \n",
    "#             final_labels_testing = np.append(final_labels_testing, labels[item][testing_true_idx], axis=0)\n",
    "#             final_labels_testing = np.append(final_labels_testing, labels[item][testing_false_idx], axis=0)\n",
    "        \n",
    "#     idx+=1\n",
    "\n",
    "# #Shuffle indeces\n",
    "# training_idx = np.random.permutation(len(final_labels_training))\n",
    "# chr_seq_training = chr_seq_training[training_idx]\n",
    "# final_labels_training = final_labels_training[training_idx]\n",
    "\n",
    "# testing_idx = np.random.permutation(len(final_labels_testing))\n",
    "# chr_seq_testing = chr_seq_testing[testing_idx]\n",
    "# final_labels_testing = final_labels_testing[testing_idx]\n",
    "\n",
    "# if data == \"seq\":\n",
    "#     np.savez_compressed(\"../data/final_data/training{}.npz\".format(name), dna=chr_seq_training, labels=final_labels_training, dtype=np.int8)\n",
    "#     np.savez_compressed(\"../data/final_data/testing{}.npz\".format(name), dna=chr_seq_testing, labels=final_labels_testing, dtype=np.int8)\n",
    "# elif data == \"chip\":\n",
    "#     np.savez_compressed(\"../data/final_data/training{}.npz\".format(name), chip=chr_seq_training, labels=final_labels_training, dtype=np.float32)\n",
    "#     np.savez_compressed(\"../data/final_data/testing{}.npz\".format(name), chip=chr_seq_testing, labels=final_labels_testing, dtype=np.float32)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b909f037",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
